### PR TITLE
Drop and recreate matview in data refresh

### PR DESCRIPTION
**Pull request by stacimc**
## Description

This PR updates the data refresh DAGs to drop and recreate the materialized view, rather than refreshing it. This is done because refreshing the view is timing out after 21 days in production, but dropping and recreating succeeds in ~5 hours.

This change is _temporary_. #1894 describes a plan for decoupling the popularity steps from the data refresh: as part of that work, these steps will be removed entirely from the data refresh DAG and refactored. However these changes will allow us to keep the image data refresh enabled while that work continues.

The data refresh DAG (note that `recreate_matview` is now a TaskGroup):
Screen Shot 2023-05-03 at 12 07 11 PM

The taskgroup expanded:
Screen Shot 2023-05-03 at 12 07 56 PM

## Testing Instructions

Try running the image and audio data refreshes locally and ensure they pass. Run each twice; assuming you have not run the refresh previously this month, the first run should refresh the popularity metrics and constants before recreating the matview, and the second should skip these steps and proceed directly to recreating the matview.

## Checklist

- [ ] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [ ] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: catalog/dags/data_refresh/dag_factory.py [hunk 1]
```diff
@@ -51,23 +51,27 @@
...
 from data_refresh.reporting import report_record_difference, report_status
 
 
 logger = logging.getLogger(__name__)
 
-REFRESH_MATERIALIZED_VIEW_TASK_ID = UPDATE_DB_VIEW_TASK_ID
+# The first task in the recreate_matview TaskGroup
+RECREATE_MATERIALIZED_VIEW_TASK_ID = (
+    f"{RECREATE_MATVIEW_GROUP_ID}" f".{DROP_DB_VIEW_TASK_ID}"
+)
```
Comment (AetherUnbound) [inline catalog/dags/data_refresh/dag_factory.py:74]: This looks a little wonky, I think black might have gotten the better of it?
```suggestion
RECREATE_MATERIALIZED_VIEW_TASK_ID = (
    f"{RECREATE_MATVIEW_GROUP_ID}.{DROP_DB_VIEW_TASK_ID}"
)
```

File: catalog/dags/data_refresh/dag_factory.py [hunk 2]
```diff
@@ -228,9 +232,10 @@ def create_data_refresh_dag(data_refresh: DataRefresh, external_dag_ids: Sequenc
             data_refresh
         )
 
-        # Refresh the materialized view. This occurs on all DagRuns and updates
-        # popularity data for newly ingested records.
-        refresh_matview = create_refresh_view_data_task(data_refresh)
+        # Drop and recreate the materialized view. This occurs on all DagRuns and
+        # updates popularity data for newly ingested records. Formerly, we would refresh
+        # the materialized view, but this process would consistently time out.
```
Comment (AetherUnbound) [inline catalog/dags/data_refresh/dag_factory.py:237]: Nit for a reword:
```suggestion
        # updates popularity data for newly ingested records. Formerly, we would refresh
        # the materialized view, but that process would consistently time out.
```

File: catalog/dags/data_refresh/recreate_view_data_task_factory.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
+"""
+# Refresh Materialized View Task Factory
```
Comment (AetherUnbound) [inline catalog/dags/data_refresh/recreate_view_data_task_factory.py:2]:
```suggestion
# Recreate Materialized View Task Factory
```

File: catalog/dags/data_refresh/recreate_view_data_task_factory.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+This file generates a Task that recreates the materialized view for a
```
Comment (AetherUnbound) [inline catalog/dags/data_refresh/recreate_view_data_task_factory.py:3]: Wouldn't this be a `TaskGroup`?
```suggestion
This file generates a TaskGroup that recreates the materialized view for a
```

File: catalog/dags/data_refresh/recreate_view_data_task_factory.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+
+GROUP_ID = "recreate_matview"
+DROP_DB_VIEW_TASK_ID = "drop_materialized_popularity_view"
+CREATE_DB_VIEW_TASK_ID = "create_materialized_popularity_view"
+
+
+def create_recreate_view_data_task(data_refresh: DataRefresh):
+    """
+    Create the refresh related tasks.
```
Comment (AetherUnbound) [inline catalog/dags/data_refresh/recreate_view_data_task_factory.py:30]:
```suggestion
    Create the recreate related tasks.
```

**Review by AetherUnbound — APPROVED** [review]
This looks great! I was able to run both of the data refresh DAGs several times locally. I have a few documentation/format corrections but nothing to block a merge.

File: catalog/dags/data_refresh/dag_factory.py (hunk 1 above)
Comment (stacimc) [reply catalog/dags/data_refresh/dag_factory.py:74]: That is very weird, yep I didn't notice it had changed the formatting. Thanks!

**Comment by stacimc** [comment]
Thank you for catching the documentation fixes, @AetherUnbound!

**Review by zackkrida — APPROVED** [review]
Awesome! So grateful this unblocks data refreshes for now while we work on further optimizations.

**Pull request merged**